File: rust/src/codex_cli.rs
use std::time::Duration;
use tokio::time::sleep;

pub struct CodexCliAgent {
    pub sandbox_mode: String,
    pub reasoning_mode: bool,
}

impl CodexCliAgent {
    pub fn new() -> Self {
        Self {
            sandbox_mode: "workspace-write".to_string(),
            reasoning_mode: false,
        }
    }

    pub fn enable_o1_reasoning(&mut self, enabled: bool) {
        self.reasoning_mode = enabled;
        println!("O1 Reasoning mode set to: {}", enabled);
    }

    pub fn set_sandbox_mode(&mut self, mode: &str) {
        self.sandbox_mode = mode.to_string();
        println!("Sandbox mode set to: {}", mode);
    }

    pub async fn request_user_approval(&self, action: &str) -> Result<bool, String> {
        println!("[TUI Prompt] User approval required for: {}", action);
        sleep(Duration::from_millis(300)).await;

        if self.sandbox_mode == "read-only" {
            println!("[TUI] Action denied by read-only sandbox");
            return Ok(false);
        }

        println!("[TUI] Action approved");
        Ok(true)
    }
}

File: go/internal/agent/codex_cli.go
package agent

import (
	"context"
	"fmt"
	"time"
)

// CodexCliAgent represents the Go port of the Codex CLI functionality
type CodexCliAgent struct {
	BaseAgent
	SandboxMode string
	ReasoningMode bool
}

// NewCodexCliAgent initializes a new Codex CLI agent
func NewCodexCliAgent() *CodexCliAgent {
	return &CodexCliAgent{
		SandboxMode:   "workspace-write",
		ReasoningMode: false,
	}
}

// EnableO1Reasoning toggles specific handling for O-series OpenAI models
func (a *CodexCliAgent) EnableO1Reasoning(enabled bool) {
	a.ReasoningMode = enabled
	fmt.Printf("O1 Reasoning mode set to: %v\n", enabled)
}

// SetSandboxMode configures the execution boundary (e.g. read-only)
func (a *CodexCliAgent) SetSandboxMode(mode string) {
	a.SandboxMode = mode
	fmt.Printf("Sandbox mode set to: %s\n", mode)
}

// RequestUserApproval simulates a TUI loop blocking for user input
func (a *CodexCliAgent) RequestUserApproval(ctx context.Context, action string) (bool, error) {
	fmt.Printf("[TUI Prompt] User approval required for: %s\n", action)
	// Simulate user considering the action
	time.Sleep(300 * time.Millisecond)
	if a.SandboxMode == "read-only" {
		fmt.Println("[TUI] Action denied by read-only sandbox")
		return false, nil
	}
	fmt.Println("[TUI] Action approved")
	return true, nil
}
